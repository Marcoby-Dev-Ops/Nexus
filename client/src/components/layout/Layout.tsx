import React, { useState } from 'react';
import Sidebar from './Sidebar';
import Header from './Header';
import PropTypes from 'prop-types';

/**
 * @interface LayoutProps
 * @description Props for the Layout component.
 * @property {React.ReactNode} children - The content to be rendered within the layout.
 * @property {Array} breadcrumbs - An array of breadcrumb objects.
 * @property {string} subtitle - The subtitle for the layout.
 */
interface LayoutProps {
  children: React.ReactNode;
  breadcrumbs?: { label: string; href?: string }[];
  subtitle?: string;
}

/**
 * @name useDarkMode
 * @description Custom hook to manage dark mode state and persistence.
 * @returns {[boolean, () => void]} - [isDark, toggleTheme]
 */
function useDarkMode(): [boolean, () => void] {
  const [isDark, setIsDark] = React.useState(() => {
    if (typeof window !== 'undefined') {
      return localStorage.getItem('theme') === 'dark' ||
        (!localStorage.getItem('theme') && window.matchMedia('(prefers-color-scheme: dark)').matches);
    }
    return false;
  });

  React.useEffect(() => {
    const root = window.document.documentElement;
    if (isDark) {
      root.classList.add('dark');
      localStorage.setItem('theme', 'dark');
    } else {
      root.classList.remove('dark');
      localStorage.setItem('theme', 'light');
    }
  }, [isDark]);

  const toggleTheme = React.useCallback(() => setIsDark((prev) => !prev), []);
  return [isDark, toggleTheme];
}

/**
 * @name Layout
 * @description The main application layout, including Sidebar, Header, and content area.
 * @param {LayoutProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Layout component.
 */
const Layout: React.FC<LayoutProps> = ({ children, breadcrumbs = [], subtitle }) => {
  const [isSidebarOpen, setIsSidebarOpen] = useState(false);
  const [isDark, toggleTheme] = useDarkMode();

  const toggleSidebar = () => {
    setIsSidebarOpen(!isSidebarOpen);
  };

  return (
    <div className="relative bg-background text-foreground">
      {/* Sidebar: popout overlay */}
      <Sidebar isOpen={isSidebarOpen} toggleSidebar={toggleSidebar} />
      {isSidebarOpen && (
        <div
          onClick={toggleSidebar}
          className="fixed inset-0 z-40 bg-black/50"
          aria-hidden="true"
        ></div>
      )}
      {/* Header: fixed at top */}
      <Header toggleSidebar={toggleSidebar} onToggleTheme={toggleTheme} isDark={isDark} />
      {/* Spacer for fixed header */}
      <div className="h-16" />
      {/* Main content: no top padding needed */}
      <main className="p-4 transition-all">
        {/* Navigation (breadcrumbs, subtitle) at top of main content */}
        {(breadcrumbs.length > 0 || subtitle) && (
          <div className="mb-4">
            {breadcrumbs.length > 0 && (
              <nav className="flex items-center text-sm text-muted-foreground truncate mb-1" aria-label="Breadcrumb">
                {breadcrumbs.map((crumb, idx) => (
                  <span key={crumb.label} className="flex items-center">
                    {crumb.href ? (
                      <a href={crumb.href} className="hover:underline text-primary font-medium">{crumb.label}</a>
                    ) : (
                      <span className="font-semibold text-foreground">{crumb.label}</span>
                    )}
                    {idx < breadcrumbs.length - 1 && <span className="mx-2">/</span>}
                  </span>
                ))}
              </nav>
            )}
            {subtitle && <div className="text-xs text-muted-foreground truncate">{subtitle}</div>}
          </div>
        )}
        {children}
      </main>
    </div>
  );
};

Layout.propTypes = {
  children: PropTypes.node.isRequired,
  breadcrumbs: PropTypes.array,
  subtitle: PropTypes.string,
};

export default Layout; 