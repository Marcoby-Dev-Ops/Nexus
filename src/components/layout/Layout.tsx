import React, { useState } from 'react';
import Sidebar from './Sidebar';
import Header from './Header';
import PropTypes from 'prop-types';

/**
 * @interface LayoutProps
 * @description Props for the Layout component.
 * @property {React.ReactNode} children - The content to be rendered within the layout.
 * @property {Array} breadcrumbs - An array of breadcrumb objects.
 * @property {string} subtitle - The subtitle for the layout.
 */
interface LayoutProps {
  children: React.ReactNode;
  breadcrumbs?: { label: string; href?: string }[];
  subtitle?: string;
}

/**
 * @name useDarkMode
 * @description Custom hook to manage dark mode state and persistence.
 * @returns {[boolean, () => void]} - [isDark, toggleTheme]
 */
function useDarkMode(): [boolean, () => void] {
  const [isDark, setIsDark] = React.useState(() => {
    if (typeof window !== 'undefined') {
      return localStorage.getItem('theme') === 'dark' ||
        (!localStorage.getItem('theme') && window.matchMedia('(prefers-color-scheme: dark)').matches);
    }
    return false;
  });

  React.useEffect(() => {
    const root = window.document.documentElement;
    if (isDark) {
      root.classList.add('dark');
      localStorage.setItem('theme', 'dark');
    } else {
      root.classList.remove('dark');
      localStorage.setItem('theme', 'light');
    }
  }, [isDark]);

  const toggleTheme = React.useCallback(() => setIsDark((prev) => !prev), []);
  return [isDark, toggleTheme];
}

/**
 * @name Layout
 * @description The main application layout, including Sidebar, Header, and content area.
 * @param {LayoutProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Layout component.
 */
const Layout: React.FC<LayoutProps> = ({ children, breadcrumbs = [], subtitle }) => {
  const [isSidebarOpen, setIsSidebarOpen] = useState(false);
  const [isDark, toggleTheme] = useDarkMode();

  const toggleSidebar = () => {
    setIsSidebarOpen(!isSidebarOpen);
  };

  return (
    <div className="flex h-screen bg-background text-foreground overflow-x-hidden">
      {/* Sidebar (Zone 1) */}
      <Sidebar isOpen={isSidebarOpen} toggleSidebar={toggleSidebar} />
      {isSidebarOpen && (
        <div
          onClick={toggleSidebar}
          className="fixed inset-0 z-20 bg-black/50 md:hidden"
          aria-hidden="true"
        ></div>
      )}
      {/* Main content area */}
      <div className="flex-1 flex flex-col min-w-0 min-h-0 overflow-hidden ml-0 md:ml-64">
        {/* Top Bar (Zone 2) */}
        <Header toggleSidebar={toggleSidebar} onToggleTheme={toggleTheme} isDark={isDark} breadcrumbs={breadcrumbs} subtitle={subtitle} />
        {/* Main content (Zones 3-5) */}
        <main className="flex-1 min-h-0 overflow-x-auto bg-background p-2 sm:p-4">
          {/* Adjust ml-64 if sidebar width changes. This is for when sidebar is static. */}
          {/* On mobile, when sidebar is an overlay, main content should not have margin. */}
          {/* We need to conditionally apply margin or adjust layout if sidebar is an overlay vs static. */}
          {/* For now, let's adjust main content padding/margin based on screen size, assuming sidebar is static on md+ */} 
          {children}
        </main>
      </div>
    </div>
  );
};

Layout.propTypes = {
  children: PropTypes.node.isRequired,
  breadcrumbs: PropTypes.array,
  subtitle: PropTypes.string,
};

export default Layout; 