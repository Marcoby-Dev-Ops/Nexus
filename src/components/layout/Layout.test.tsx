/**
 * @file Layout.test.tsx
 * @description Unit and snapshot tests for the Layout layout component.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Layout from './Layout';

describe('Layout', () => {
  it('renders children', () => {
    render(
      <Layout>
        <div data-testid="child">Child Content</div>
      </Layout>
    );
    expect(screen.getByTestId('child')).toBeInTheDocument();
    expect(screen.getByText('Child Content')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Layout>
        <div>Snapshot Child</div>
      </Layout>
    );
    expect(asFragment()).toMatchSnapshot();
  });
}); 