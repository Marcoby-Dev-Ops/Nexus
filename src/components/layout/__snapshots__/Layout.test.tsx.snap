// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen bg-background "
  >
    <main
      class="container mx-auto px-4 py-8"
    >
      <div>
        Snapshot Child
      </div>
    </main>
  </div>
</DocumentFragment>
`;
